File: src/main/java/com/tikal/fleet/rxjava/ServiceMonitorThreaded.java
package com.tikal.fleet.rxjava;

import com.google.gson.JsonElement;
import com.google.gson.JsonParser;
import org.apache.http.HttpEntity;
import org.apache.http.HttpResponse;
import org.apache.http.HttpStatus;
import org.apache.http.client.HttpClient;
import org.apache.http.client.config.RequestConfig;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.impl.client.HttpClientBuilder;
import org.apache.http.util.EntityUtils;
import rx.Observable;
import rx.Subscription;
import rx.schedulers.Schedulers;

import java.io.IOException;

/**
 * Created by chaimturkel on 6/29/16.
 */
public class ServiceMonitorThreaded {



    public void getServicesStatus(){
        final JsonParser jsonParser = new JsonParser();

        Observable<String> urlRequest1 = Observable.just("http://jsonplaceholder.typicode.com/posts/1");
        Observable<String> urlRequest2 = Observable.just("http://jsonplaceholder.typicode.com/posts/2");
        Observable<String> urlRequest3 = Observable.just("http://jsonplaceholder.typicode.com/error/2");

        Observable<String> allRequests = Observable.concat(urlRequest1, urlRequest2,urlRequest3);
        allRequests.flatMap(request -> Observable.just(request)
                .map(url -> Utils.getResponse(url))
                .filter(response -> response.getStatusLine().getStatusCode() == HttpStatus.SC_OK)
                .map(response -> jsonParser.parse(Utils.httpEntityToString(response.getEntity())))
                .subscribeOn(Schedulers.computation())
        ).toList()
        .subscribe(
                jsonElementList -> {
                    for (JsonElement jsonElement : jsonElementList) {
                        System.out.println("element - " + jsonElement.toString());
                    }
                },
                error -> System.out.println("error"),
                () -> System.out.println("completed"));


        try {
            Thread.sleep(20000);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

    }

    public static void main(String[] args) {
        ServiceMonitorThreaded serviceMonitor = new ServiceMonitorThreaded();
        serviceMonitor.getServicesStatus();
    }

}

File: src/main/scala/com/tikal/common/futures/ServiceMonitorThreaded.scala
package com.tikal.common.futures

import com.google.gson.{JsonElement, JsonParser}
import com.tikal.fleet.rxjava.Utils
import org.apache.http.{HttpResponse, HttpStatus}

import scala.concurrent.Future
import scala.util.{Failure, Success}

/**
  * Created by chaimturkel on 6/30/16.
  */
object ServiceMonitorThreaded extends App{

  import scala.concurrent.ExecutionContext.Implicits._
  val urls = List("http://jsonplaceholder.typicode.com/posts/1",
    "http://jsonplaceholder.typicode.com/posts/2",
    "http://jsonplaceholder.typicode.com/error/3")
  val jsonParser = new JsonParser

  var listFuture : List[Future[Option[JsonElement]]] =
    for (url <- urls)
      yield Future { Utils.getResponse(url) }
        .map(http => {
        if (http.getStatusLine.getStatusCode == HttpStatus.SC_OK)
          Some(jsonParser.parse(Utils.httpEntityToString(http.getEntity())))
        else None
      })

  val list : Future[List[Option[JsonElement]]] = Future.sequence(listFuture)
  list onComplete {
    case Success(posts) => for (post <- posts) if (post.isDefined) println(post.toString)
    case Failure(t) => println("An error has occurred: " + t.getMessage)
  }

  Thread.sleep(100000)
}

